File: PermSearch.cpp
#include <iostream>
#include <array>
#include <unordered_map>
#include <vector>
#include <boost/multiprecision/miller_rabin.hpp>


using Array = std::array<int, 14>;
namespace std {
    template <>
    struct hash<std::array<int, 14>> {
        uint64_t operator()(const Array& key) const {
            uint64_t x = 0;
            for(int i : key) {
                x ^= i;
                x ^= (x << 13);
                x ^= (x >> 7);
                x ^= (x << 17);
            }
            return x;
        }
    };
}

struct DFSTree {
    using BigInt = boost::multiprecision::cpp_int;
    DFSTree* child[10] = {};
    BigInt num;
    int n, eleven;
    std::unordered_map<Array, int> cards;
    DFSTree(BigInt num = 0): num(num) {}
    DFSTree& at(int x) {
        if(child[x] == nullptr) {
            child[x] = new DFSTree(num * 10 + x);
            child[x]->n = n - 1;
            child[x]->eleven = x - eleven;
        }
        return *child[x];
    }
    bool is_valid(const Array& s, int x) {
        if(n == 0) return eleven % 11;
        if(x % 2 && x != 5) if(!s[1] && !s[3] && !s[7] && !s[9] && !s[11] && !s[13]) return false;
        if(x < 10) if(!s[0] && !s[1] && !s[2] && !s[3] && !s[4] && !s[5] && !s[6] && !s[7] && !s[8] && !s[9])
            if((eleven - s[10] + s[12] + s[13] * 2) % 11 == 0) return false;
        return true;
    }
    void insert(const Array& s, int x) {
        Array t = s;
        t[x]--;
        DFSTree& next = x < 10 ? at(x) : at(1).at(x - 10);
        if(next.is_valid(t, x)){
            if(!next.cards.count(t)) next.cards[t] = x;
            else next.cards[t] = std::min(next.cards[t], x);
        }
    }
    void search(std::vector<BigInt>& ans, int cnt = 1) {
        if(n == 0) {
            if(cards.size() && boost::multiprecision::miller_rabin_test(num, 50)) ans.push_back(num);
            return;
        }
        for(auto& i : cards) {
            if(i.second == 1) {
                for(int j = 0; j < 4; j++) if(i.first[j] && !i.first[j + 10]) insert(i.first, j);
                for(int j = 4; j < 11; j++) if(i.first[j]) insert(i.first, j);
                for(int j = 12; j < 14; j++) if(i.first[j]) insert(i.first, j);
            }
            else {
                for(int j = 0; j < 14; j++) if(i.first[j]) insert(i.first, j);
            }
        }
        for(int i = 9; i >= 0; i--) if(child[i] != nullptr) {
            if(ans.size() != cnt) child[i]->search(ans, cnt);
            delete child[i];
            child[i] = nullptr;
        }
    }
};
std::vector<boost::multiprecision::cpp_int> PermSearch(const std::vector<int>& hand, int n = 1) {
    if(hand.size() > 1 && accumulate(hand.begin(), hand.end(), 0) % 3 == 0) return {};
    Array cards;
    cards.fill(0);
    for(auto& i : hand) cards[i]++;
    DFSTree a;
    a.cards[cards] = 0;
    a.n = 0;
    for(int i = 0; i < 10; i++) a.n += cards[i];
    for(int i = 10; i < 14; i++) a.n += cards[i] * 2;
    std::vector<boost::multiprecision::cpp_int> ans;
    a.search(ans, n);
    return ans;
}

// 使用例

int main(){
    std::vector<int> hand = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13};
    for(auto& i : PermSearch(hand, 10)) std::cout << i << std::endl;
    /*
     98765432131121011
     98765432121311101
     98765432121101311
     98765432121011311
     98765432111210113
     98765432111012113
     98765432111011213
     98765432110111213
     98765432101311121
     98765431321101211
    */

   hand = {1, 7, 7, 9, 13};
   for(auto& i : PermSearch(hand, -1)) std::cout << i << std::endl;
   /*
    971371
    917713
    913771
    913177
    797131
    791317
    779131
    719713
    717139
    711397
    197713
    177913
    139177
    137791
    137197
    131779
    113797
   */
}

File: PermSearch.py
p = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97]
def suspect(a, s, d, n):
    x = pow(a, d, n)
    if x == 1:
        return True
    for r in range(s):
        if x == n - 1:
            return True
        x = x * x % n
    return False

def miller_rabin_test(n):
    if n <= 1:
        return False
    if n <= 100:
        return n in p
    if (n & 1) == 0:
        return False
    d = n - 1
    s = 0
    while (d & 1) == 0:
        s += 1
        d >>= 1
    for i in p:
        if not suspect(i, s, d, n):
            return False
    return True

class DFSTree:
    def __init__(self, num = 0):
        self.child = [None] * 10
        self.num = num
        self.n = 0
        self.eleven = 0
        self.cards = dict()
    def at(self, x): # 1桁進める
        if self.child[x] == None:
            self.child[x] = DFSTree(self.num * 10 + x)
            self.child[x].n = self.n - 1
            self.child[x].eleven = x - self.eleven
        return self.child[x]
    def is_valid(self, s, x): # 2, 5, 11 の倍数チェック
        if self.n == 0:
            return self.eleven % 11
        if x % 2 and x != 5:
            if not (s[1] or s[3] or s[7] or s[9] or s[11] or s[13]):
                return False
        if x < 10:
            if not (s[0] or s[1] or s[2] or s[3] or s[4] or s[5] or s[6] or s[7] or s[8] or s[9]):
                if (self.eleven - s[10] + s[12] + s[13] * 2) % 11 == 0:
                    return False
        return True
    def insert(self, s, x): # s : 残っているカード, x: 次に使うカード
        t = list(s)
        t[x] -= 1
        nxt = self.at(x) if x < 10 else self.at(1).at(x - 10)
        if nxt.is_valid(t, x):
            t = tuple(t)
            if t not in nxt.cards:
                nxt.cards[t] = 14
            if nxt.cards[t] > x:
                nxt.cards[t] = x
    def search(self, ans, cnt = 1): # ans : 0~Kの枚数の配列, cnt : 何個探すか(-1で全部)
        if self.n == 0:
            if self.cards and miller_rabin_test(self.num):
                ans.append(self.num)
            return
        for s, last in self.cards.items():
            if last == 1:
                for i in range(4):
                    if s[i] and not s[i + 10]:
                        self.insert(s, i)
                for i in range(4, 11):
                    if s[i]:
                        self.insert(s, i)
                for i in range(12, 14):
                    if s[i]:
                        self.insert(s, i)
            else:
                for i in range(14):
                    if s[i]:
                        self.insert(s, i)
        for i in reversed(range(10)):
            if self.child[i] != None:
                if len(ans) != cnt:
                    self.child[i].search(ans, cnt)
                self.child[i] = None # destruct

def PermSearch(hand, n = 1):  # -> list<Int>
    if len(hand) > 1 and sum(hand) % 3 == 0:
        return []
    cards = [0] * 14
    for i in hand:
        cards[i] += 1
    a = DFSTree()
    a.cards[tuple(cards)] = 0
    a.n = 0
    for i in range(10):
        a.n += cards[i]
    for i in range(10, 14):
        a.n += cards[i] * 2
    ans = []
    a.search(ans, n)
    return ans

# 使用例

hands = [2, 3, 4, 5, 6]
for i in PermSearch(hands, -1):
    print(i)
'''
65423
54623
46523
45263
25643
25463
'''

hands = [i for i in range(1, 14)] * 4
hands.append(12)
hands.append(13)
for i in PermSearch(hands):
    print(i)
'''
999988887777666655554444333322221313131313121212121211111111101110110101
'''
